### ISSUE TITLE
Performance drop when enabling bias in FastDense

### ISSUE THREAD

**Issue by tkmamp** [SciML/DiffEqFlux.jl#757]
Hello there, 

I came across an interesting property (or issue) of the (Fast)Dense struct for that I also propose a solution here. 

### TL;DR
When activating the bias of the Dense Layers (tested DiffEqFlux.FastDense and Flux.Dense), that are used to enhance an ODE function, the performance of the gradient calculation via Zygote.gradient drops significantly. This effect even worses, when multiple NNs are used. 
It shows, that when calculating the output of a dense layer via `` [W|b][x|1]^T `` instead of `` Wx + b ``, the performance is stable for activated bias.

### MWE
Here a (not so minimal) working example. The ODE describes the dynamics of a two body system, connected by a spring-damper pair. One mass is excited via a second spring-damper pair, here very simple with a sinus-wave. Anyway, what the ODE calculates is not the point here, so I will only go into detail when requested. 

In this MWE, data is created by simulating the ODE with an additional non-linearity, that should be learned by two NNs during training. I added 2 NNs solely because the performance decrease shows more clearly when using multiple NNs. 

In the loss function, the ODE is simulated again (this time without the non-linearity) and the squared error for each point of the trajectory is added up and finally meaned to obtain the MSE.

This loss is then differentiated with ForwardDiff and Zygote and the performance measured via ``` @time ``` 

```julia
using OrdinaryDiffEq
using DiffEqFlux
using ReverseDiff
using ForwardDiff
using Zygote

###### model params
m_b = 450.         # mass body
m_w = 60.          # mass wheel
k_s = 30000.       # spring stiffness suspension
d_s = 900.         # damper  suspension
k_t = 280000.      # spring stiffness tire
d_t = 50.            # damping tire

# intial condition
x0 = zeros(5)

# constant parameters
p_model = [m_b, m_w, k_s, d_s, k_t, d_t]

# excitation (just a sinus wave)
u_input(t) = 0.25*sin((1*2*pi)*t)

# calculate NN outpus (usage of 2 nns makes no real sense here but the performance drop can be better shown this way)
function calc_nn_outputs(dz, p)
    nn_out_1 = nn_1([dz], p[p_idx_1])[1]
    nn_out_2 = nn_2([dz], p[p_idx_2])[1]
    return nn_out_1 + nn_out_2
end

#### ode definition
function linear_qvm_ode!(dx, x, u, p_nn, t)
    m_b, m_w, k_s, d_s, k_t, d_t = p_model
   
    dz_dot_s = x[4] - x[2]
    dz_s = x[3] - x[1]
    dz_dot_t = u(t) - x[4]
    dz_t = x[5] - x[3]

    F_delta_s = 0.
    if abs(sum(p_nn)) < 1e-12
        # adding non-linarity during data emulation
        F_delta_s = (1000*dz_s)^3 + 1000*dz_s
    else
        F_delta_s = calc_nn_outputs(dz_s, p_nn)
    end

    F_d = d_s*dz_dot_s
    F_k = k_s*dz_s
    F_d_t = d_t*dz_dot_t
    F_k_t = k_t*dz_t

    dx[1] = x[2]
    dx[2] = 1/(m_b)*(F_k + F_d + F_delta_s)
    dx[3] = x[4]
    dx[4] = 1/(m_w)*(F_k_t + F_d_t - F_k - F_d - F_delta_s)
    dx[5] = u(t)
end

function x_loss(p_opt, prob, t_ref, x_ref, x0)
    # define & solve ode-problem   
    sens = ReverseDiffAdjoint()    
    sol = solve(remake(prob; p=p_opt, u0=x0, tspan=[t_ref[1], t_ref[end]]), Tsit5(), p=p_opt, reltol=1e-4; sensealg=sens, saveat=t_ref)
    # calculate loss of the solution
    loss = 0.
    for k in 1:size(sol)[2]
        for i in 1:(size(sol)[1]-1)
            loss += sum(abs2(sol[i, k] - x_ref[i, k])) #sum(squared error)
        end
    end    
    loss *=1/(size(sol)[2]) # mean over timesteps
    return loss
end

###### define ode 
ode_function = (dx, x, p, t) -> linear_qvm_ode!(dx, x, u_input, p, t)

###### emulate data: simulate for tspan with added non-linarities
t_start = .0
t_end = 10.
tspan = [t_start, t_end]

dt_sol = 0.01
t_sol = t_start:dt_sol:t_end

prob = ODEProblem(ode_function, x0, tspan)
x_ref = solve(prob, Tsit5(), p=[0.], reltol=1e-4, saveat=t_sol)


######## define loss for gradient ###########
loss(p_opt) = x_loss(p_opt, prob, t_sol, x_ref, x0)


##########################################################################
############# Comparison DiffEqFlux Fast Chain vs. MyChain ###############

######## FastChain first
n_units_base = 20

println("________________________")
println("DiffEqFlux: FastChain")
for i in 1:2
    n_units = n_units_base-(i==2)*1 # approximately the same number of parameters for bias on/off
    bias_on = i==2

    global nn_1 = FastChain(FastDense(1, n_units, relu, bias=bias_on), 
                          FastDense(n_units, n_units, relu, bias=bias_on), 
                          FastDense(n_units, n_units, relu, bias=bias_on), 
                          FastDense(n_units, 1, identity, bias=bias_on))
    
    global nn_2 = FastChain(FastDense(1, n_units, relu, bias=bias_on), 
                          FastDense(n_units, n_units, relu, bias=bias_on), 
                          FastDense(n_units, n_units, relu, bias=bias_on), 
                          FastDense(n_units, 1, identity, bias=bias_on))

    p_nn_1 = initial_params(nn_1)
    p_nn_2 = initial_params(nn_2)
    
    global p_idx_1 = 1:length(p_nn_1)
    global p_idx_2 = length(p_nn_1)+1:length(p_nn_1)+length(p_nn_2)

    p_train = vcat(p_nn_1, p_nn_2)

    if i == 1
        println("")  
        println("bias off")  
    else
        println("")
        println("bias on")
    end             
    
    println(string("#parameters: ", length(p_train)))

    print("ForwardDiff.gradient: ")
    ForwardDiff.gradient(loss, p_train)
    @time ForwardDiff.gradient(loss, p_train)
   
    print("Zygote.gradient:      ")  
    Zygote.gradient(loss, p_train)
    @time Zygote.gradient(loss, p_train)
end


################################################
######### MyChain definition
struct MyDense
    in::Int
    out::Int
    bias::Bool
    activation::Function
    init_params::Function
    function MyDense(in::Int, out::Int; activation=identity, bias=true)            
        init_params() = ((bias == false) ? vec(Flux.glorot_uniform(out, in)) : vcat(vec(Flux.glorot_uniform(out, in)), Flux.zeros32(out)))
        new(in, out, bias, activation, init_params)
    end
end

# New method [W|b]*[x|1]' (basically copy - paste - adapt of the FastDense function
(f::MyDense)(x::AbstractVector, p) = ((f.bias == true) ? (f.activation.((reshape(p, f.out, f.in+1))*vcat(x, 1))) : (f.activation.(reshape(p[1:(f.out*f.in)], f.out, f.in)*x)))

struct MyChain
    layers::Vector{MyDense}
    p_idx::Vector{UnitRange}
    p_init::Vector{Float32}
    function MyChain(layers::Vector{MyDense})
        p_idx = []
        p_init = Vector{Float32}([])
        k = 0
        for i in 1:length(layers)
            l_i = layers[i]
            push!(p_idx, k+1:k+length(l_i.init_params()))
            k += length(l_i.init_params())
            p_init = vcat(p_init, l_i.init_params())
        end
        new(layers, p_idx, p_init)
    end
end

# MyChain function call
function (f::MyChain)(x::AbstractVector, p)
    y = f.layers[1](x, p[f.p_idx[1]])
    for j = 2:length(f.layers)-1
        y = f.layers[j](y, p[f.p_idx[j]])
    end
    return f.layers[end](y, p[f.p_idx[end]])
end

#########################################
########### Time MyChain
println("")
println("________________________")
println("MyChain")
for i in 1:2
    n_units = n_units_base-(i==2)*1 # approximately the same number of parameters for bias on/off
    bias_on = i==2

    layers = [MyDense(1, n_units, activation=relu, bias=bias_on), 
              MyDense(n_units, n_units, activation=relu, bias=bias_on), 
              MyDense(n_units, n_units, activation=relu, bias=bias_on), 
              MyDense(n_units, 1, activation=identity, bias=bias_on)]

    global nn_1 = MyChain(layers)
    global nn_2 = MyChain(layers)

    p_nn_1 = nn_1.p_init
    p_nn_2 = nn_2.p_init
    
    global p_idx_1 = 1:length(p_nn_1)
    global p_idx_2 = length(p_nn_1)+1:length(p_nn_1)+length(p_nn_2)

    p_train = vcat(p_nn_1, p_nn_2)

    if i == 1
        println("")  
        println("bias off")  
    else
        println("")
        println("bias on")
    end             
    
    println(string("#parameters: ", length(p_train)))

    print("ForwardDiff.gradient: ")
    ForwardDiff.gradient(loss, p_train)
    @time ForwardDiff.gradient(loss, p_train)
    
    print("Zygote.gradient:      ")  
    Zygote.gradient(loss, p_train)
    @time Zygote.gradient(loss, p_train)
end
```

![mwe_bias_timingsnip](https://user-images.githubusercontent.com/108574207/183929453-e4c808d9-4c2e-4d0c-be40-8102e16fc21d.png)

In the Snip you can see, that Zygote outperforms ForwardDiff as expected (blue) when the bias is deactivated. When the bias is activated, it is still faster but now roughly 6 times slower than without the bias (and even a few parameters less, red marked). 

### Proposed Solution
When discussing this with fellow developers, they hinted that when evaluating a NN, it is advantageous to calculate the output of a dense layer as `` [W|b][x|1]^T `` instead of `` Wx + b `` . I implemented that in MyDense and MyChain and the performance is now stable for activated bias (see output picture, green marked). 

Question: Was the implementation this way on purpose? The vcat(x, 1) call might slightly decrease performance in the (forward) evaluation of the Dense Layer. Anyway I think this is neglectable when calculating gradients through integration over longer timespans. 

### Appendix
Tested in Julia 1.7.2
DiffEqFlux = "aae7a2af-3d4f-5e19-a356-7da93b79d9d0"
OrdinaryDiffEq = "1dea7af3-3e70-54e6-95c3-0bf5283fa5ed"
ForwardDiff = "f6369f11-7733-5829-9624-2563aa707210"
ReverseDiff = "37e2e3b7-166d-5795-8a7a-e32c996b4267"
Zygote = "e88e6eb3-aa80-5325-afca-941959d7151f"

**Comment by ChrisRackauckas** [comment]
FastChain was removed a few months ago. Test Lux.jl.

**Comment by tkmamp** [comment]
Dang, ok. Thank you.

**Comment by ChrisRackauckas** [comment]
Yeah no problem. Yeah FastChain had two concepts: pre-allocated reverse mode buffers, and fully function interfaces. Lux is the latter, SimpleChains is the former. Those are now being maintained as full neural network libraries. So this one has no purpose, and is now filled with deprecation warnings that would slow it down. It's being deleted in the next breaking release but being left around so people can see the deprecation warnings.

**Comment by tkmamp** [comment]
Good to know!
At a first glance, SimpleChains should be (better) suited for this kind of application so I'll try that first 🤙
